### ISSUE TITLE
Website gets into a state where out going HttpClient connections continuously fail.

### ISSUE THREAD

**Issue by sobenz** [aspnet/Home#3234]
**TL;DR;** We had a simple service that was running without problem as a Full AspNet (v4.6.1) WebApp on Azure with a common set of business logic libaries in DotNet Standard. On migration to AspNet Core (firsty v2.0, and currently v2.1) on DotNetCore v2.0 we have been getting cases where all outgoing traffic via the HttpClient begins to fail to establish outgoing connections. Once in this state it does not seem to recover.

**Details:**
The stack trace below is an example of the types of exception we receive:

```
2018-06-16 17:12:16.947 +00:00 [Error] web-error-logger: Unhandled web exception at: GET:/api/consent/tenants/LGM695/consumers/e3f1f472-01e6-4c35-8072-355d907b08e4
Microsoft.WindowsAzure.Storage.StorageException: An error occurred while sending the request. ---> System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.Net.Http.WinHttpException: A connection with the server could not be established
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Threading.Tasks.RendezvousAwaitable`1.GetResult()
   at System.Net.Http.WinHttpHandler.<StartRequest>d__105.MoveNext()
   --- End of inner exception stack trace ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpClient.<FinishSendAsyncUnbuffered>d__59.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at Microsoft.WindowsAzure.Storage.Core.Executor.Executor.<ExecuteAsyncInternal>d__4`1.MoveNext()
   --- End of inner exception stack trace ---
   at Microsoft.WindowsAzure.Storage.Core.Executor.Executor.<ExecuteAsyncInternal>d__4`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at *****.Gdpr.Consent.Storage.ConsentInfoTableStorage.<GetCurrentConsentAsync>d__10.MoveNext() in D:\a\1\s\src\*****.Gdpr.Consent\Storage\ConsentInfoTableStorage.cs:line 105
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at *****.Gdpr.Consent.Business.ConsentBusiness.<SearchForConsumerConsentAsync>d__14.MoveNext() in D:\a\1\s\src\*****.Gdpr.Consent\Business\ConsentBusiness.cs:line 217
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at *****.Gdpr.Consent.Service.Controllers.ConsentController.<GetConsumerConsentAsync>d__4.MoveNext() in D:\a\1\s\src\*****.Gdpr.Consent.Service\Controllers\ConsentController.cs:line 61
```

- Business & Storage layers of code are unchanged between the functional ASP.Net version and ASP.Net Core versions, and controller code is very liteweight (example below)

```
public async Task<IActionResult> GetConsumerConsentAsync([FromRoute]string tenantId, [FromRoute]Guid consumerId)
{
    if ((await GetConfiguredTenant(tenantId)) == null)
        return StatusCode(403, new ErrorInfo { StatusCode = 403, Message = StringResources.ErrorTenantNotConfigured, Details = JToken.FromObject(tenantId) });
    var consentOperation = new ConsentOperation { TenantId = tenantId, ConsumerId = consumerId };
    ConsentInfo consentInfo = await _consentBiz.SearchForConsumerConsentAsync(consentOperation);
    if (consentInfo == null)
        return NotFound();
    return Ok(consentInfo);
}
```

- Failures are not limited to Azure Storage as we have seen similar failures where the JWT middleware component fails to communicate with AzureA.
- Current libraries are all on the latest no preview version of AspNet Core / Azure Storage etc. The application itself is running currently as .NetCore 2.0 rather than 2.1
- Flow is async all the way through from controller -> business -> storage
- Replication is difficult as site can run fine for several days with no obvious issues (i.e. cpu/memory/thread count roughly static and not gradually increasing) at a load of about 200->3000 rpm depending on time of day. Load does appear however to be somewhat related as failures do tend to be during the bussier times.

I have read that in dotnet core 2.1 has a complete re-write of the underlying Http network connectivity used by HttpClient. Is anyone aware of the dotnet core 2.0 handling of http calls on windows causing these issues?

**Comment by khellang** [comment]
Sounds like socket exhaustion. Can you show your usage of `HttpClient`?

**Comment by sobenz** [comment]
@khellang - I agree it does sound like a socket related issue. We don't use a HttpClient directly, the service primarily makes calls to Azure Table Storage via the standard MS library. All the operations are generally straight forward such as:
```
var operation = TableOperation.Retrieve<ConsentEntity>(consumerId.ToString(), string.Empty);
var tableResult = await table.ExecuteAsync(operation).ConfigureAwait(false);
```
Looking at the TCP connections for the Azure WebApp there is a spike of out going TCP connections around the failure, but these seem to be dwarfed by the loopback connections (which I am guessing are the W3WP -> Kestrel connections). 
![image](https://user-images.githubusercontent.com/11184586/41558859-5539a6f4-7396-11e8-89bb-ec0b3392ec2f.png)

**Comment by sobenz** [comment]
It appears that this issue is environmental as issue cannot be replicated when on a different App Service Plan. Addressing the issue with Azure Support.
